### DISCUSSION TITLE
Testing the nested lookahead stuff

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#143] [Q&A]
Are there any existing tests for the nested lookahead stuff? With this grammar adapted from your post:
```
PARSER_PACKAGE=nla_test;
TREE_BUILDING_ENABLED=false;
BASE_NAME="Test";

INJECT TestParser :
    import java.nio.file.Paths;
{
    public String lookahead_status;
    public String parse_status;

    public static void main(String args[]) throws IOException {
        TestParser p = new TestParser(Paths.get(args[0]));
        try {
            p.Root();
            p.parse_status = "parse succeeded";
        }
        catch (ParseException pe) {
            p.parse_status = "parse failed";
        }
        System.out.println(String.format("%s, %s", p.lookahead_status, p.parse_status));
    }
}

SKIP : " " ;

NestedChoice :
   SCAN 2 "foo" "bar" "baz"
   |
   "foo"
;

Root :
    SCAN NestedChoice => {lookahead_status = "lookahead succeeded";} NestedChoice
    |
    {lookahead_status = "lookahead failed";}
;
```
and this test script (where `ccc` is basically `java -jar congocc.jar`):
```sh
ccc -q NLA.ccc
javac nla_test/TestParser.java
echo Trying good ...
java nla_test.TestParser nla-good.txt
echo Trying bad ...
java nla_test.TestParser nla-bad.txt
echo Rebuilding with legacy glitchy lookahead ...
ccc -p LEGACY_GLITCHY_LOOKAHEAD=true -q NLA.ccc
javac nla_test/TestParser.java
echo Trying good ...
java nla_test.TestParser nla-good.txt
echo Trying bad ...
java nla_test.TestParser nla-bad.txt
```
and these test files, `nla-good.txt` first:
```
foo bar baz
```
 and `nla-bad.txt` after:
```
foo bar foo
```
The script, when run, prints:
```console
Trying good ...
lookahead succeeded, parse succeeded
Trying bad ...
lookahead failed, parse succeeded
Rebuilding with legacy glitchy lookahead ...
Trying good ...
lookahead succeeded, parse succeeded
Trying bad ...
lookahead succeeded, parse failed
```
Is this as expected, and is this a sufficient test for nested lookahead? Feel free to post code/data snippets any additional things you think should be tested.

**Comment by revusky** [comment]
> Is this as expected, 

Yes, it seems that this is working as it should. Without `LEGACY_GLITCHY_LOOKAHEAD` enabled, the lookahead fails because, even though it checked for the `foo` `bar` `baz`, and that failed, it does *not* go on to the next choice, the lone `foo`, the reason being that it already passed the two-token "predicate", so the correct logic is *not* to go on to the next choice. You understand the issue, right? The lookahead code should reflect (to the maximum extent possible surely) the same logic as the parser code. If you're parsing, and you pass those two tokens, you are *committed* to that branch. So you don't go on to later choices. *It doesn't actually matter if one of those later choices would have succeeded*. (This is the tool's core logic which could be confusing for a new user, but that's how the thing works.) So, again, the lookahead logic should reflect that, but with `LEGACY_GLITCHY_LOOKAHEAD` enabled (replicating the earlier incorrect behavior) the lookahead succeeds, but the parsing then fails. (Of course, a new user, who is writing completely new grammars from scratch and never used an earlier version of the tool where this bug was present, would be better off not even knowing about the `LEGACY_GLITCHY_LOOKAHEAD` setting!)

> and is this a sufficient test for nested lookahead?

Well, it seems that this test does get to the heart of the issue. Whether it's truly sufficient or not... well... I guess I'd say that it's good enough for now. I mean, there are so many things to turn our attention to that...  Well, come to think of it, maybe another similar test that, rather than use `SCAN`, expresses the same basic thing with an up-to-here. That could be good, I guess, and then I guess we're okay. Probably better to get back to other things, and if there are other issues with this somehow, it will eventually come to our attention.

Do we now have this working correctly on the Python and/or C# end?

**Comment by vsajip** [comment]
> Do we now have this working correctly on the Python and/or C# end?

Probably not, but writing tests will help to identify that and act as a sanity check when templates change. I have updated the templates and ensured that all current tests pass, so that's something - but I've probably missed one or two places.

**Comment by vsajip** [comment]
OK, I've done the tests and they pass locally (Java, Python, C#), but on GitHub CI I'm having some issues with .NET versions. For now, I'll disable the C# tests while I look into it.

**Comment by vsajip** [comment]
Resolved the .NET version issues. All done :smile:
